File: Data_Structures/python/05_키로거.py
# fast campus 강의
# https://www.acmicpc.net/problem/5397
# 1

test_case = int(input())

for _ in range(0, test_case):
    s = input()
    left_stack = []
    right_stack = []
    for c in s:
        if c == "<" and left_stack:
            right_stack.append(left_stack.pop())
        elif c == ">" and right_stack:
            left_stack.append(right_stack.pop())
        elif c == "-" and left_stack:
            left_stack.pop()
        elif c != "<" and c != ">" and c != "-":
            left_stack.append(c)

    left_stack.extend(reversed(right_stack))
    print(''.join(left_stack))

File: Data_Structures/c++/05_키로거.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/5397
// 1
#include <iostream>
#include <stack>
#include <vector>
#include <algorithm>

using namespace std;

int main()
{
	int test_case;
	cin >> test_case;

	for (int t = 0; t < test_case; t++) {
		string s;
		stack<char> front;
		stack<char> back;
		cin >> s;

		for (int i = 0; i < s.size(); i++) {
			if (s[i] == '<' && !front.empty()) {
				back.push(front.top());
				front.pop();
			}
			else if (s[i] == '>' && !back.empty()) {
				front.push(back.top());
				back.pop();
			}
			else if (s[i] == '-' && !front.empty())
				front.pop();
			else
				front.push(s[i]);
		}

		vector<char> result;
		while (!back.empty()) {
			front.push(back.top());
			back.pop();
		}
		while (!front.empty()) {
			result.push_back(front.top());
			front.pop();
		}
		
		reverse(result.begin(), result.end());
		for (auto i : result)
			cout << i;
		cout << '\n';
	}
}